dartagnan/src/main/java/com/dat3m/dartagnan/verification/model/ExecutionModelManager.java

```diff
@@ -0,0 +1,669 @@
+                    relGraphCache.put(r, recGraph);
+                }
+            }
+            for (DependencyGraph<Relation>.Node node : component) {
+                Relation r = node.getContent();
+                RelationGraph rg = createGraph(r);
+                if (r.isRecursive()) {
+                    RecursiveGraph recGraph = (RecursiveGraph) relGraphCache.get(r);
+                    recGraph.setConcreteGraph(rg);
+                } else {
+                    relGraphs.add(rg);
+                }
+            }
+        }
+
+        Set<CAATPredicate> predicates = new HashSet<>(relGraphs);
+        PredicateHierarchy hierarchy = new PredicateHierarchy(predicates);
+        hierarchy.initializeToDomain(domain);
+
+        // Populate graphs of base relations.
+        for (CAATPredicate basePred : hierarchy.getBasePredicates()) {
+            Relation r = relGraphCache.inverse().get((RelationGraph) basePred);
+            try {
+                r.getDefinition().accept(graphPopulator);
+            } catch (UnsupportedOperationException e) {
+                // Populate graph of relations unsupported by the visitor using default relation analysis.
+                graphPopulator.populateDynamicDefaultGraph(r);
+            }
+        }
+
+        // Do the computation.
+        hierarchy.populate();
+
+        for (CAATPredicate pred : hierarchy.getPredicateList()) {
+            Relation r = relGraphCache.inverse().get((RelationGraph) pred);
+            if (relModelCache.containsKey(r)) {
+                RelationModel rm = relModelCache.get(r);
+                ((RelationGraph) pred).edgeStream()
+                                      .forEach(e -> rm.addEdgeModel(getOrCreateEdgeModel(e)));
+           }
+        }
+
+        for (Relation r : relsToExtract) {
+            executionModel.addRelation(r, relModelCache.get(r));
+        }
+    }
+
+    private void createModel(Relation r) {
+        if (relModelCache.containsKey(r)) { return; }
+        relModelCache.put(r, new RelationModel(r));
+    }
+
+    private RelationGraph createGraph(Relation r) {
+        RelationGraph rg;
+        try {
+            rg = r.getDefinition().accept(graphBuilder);
+        } catch (UnsupportedOperationException e) {
+            // Generate a SimpleGraph for base relations that can be populated manually.
+            rg = new SimpleGraph();
+        }
+        rg.setName(r.getNameOrTerm());
+        if (!r.isRecursive()) {
+            relGraphCache.put(r, rg);
+        }
+        return rg;
+    }
+
+    private RelationGraph getOrCreateGraph(Relation r) {
+        if (relGraphCache.containsKey(r)) {
+            return relGraphCache.get(r);
+        }
+        return createGraph(r);
+    }
+
+    private EdgeModel getOrCreateEdgeModel(Edge e) {
+        String identifier = e.getFirst() + " -> " + e.getSecond();
+        if (edgeModelCache.containsKey(identifier)) {
+            return edgeModelCache.get(identifier);
+        }
```

Comment:
Why do you need to make strings out of the edge? Can't you just use the `Edge` as key in your cache?